pkg/grizzly/workflow.go

```diff
@@ -96,93 +100,95 @@ func eval(jsonnetFile string) ([]byte, error) {
 	}
 
-	// Destructure Mixin into Resources slice
-	var r Resources
-	for _, b := range m.Dashboards {
-		r = append(r, b)
-	}
-
-	for _, g := range m.Rules.Groups {
-		r = append(r, g)
-	}
-
-	for _, g := range m.Alerts.Groups {
-		r = append(r, g)
+	r := Resources{}
+	for k, v := range msi {
+		handler, err := config.Registry.GetHandler(k)
+		if err != nil {
+			fmt.Println("Skipping unregistered path", k)
+			continue
+		}
+		resources, err := handler.Parse(v)
+		if err != nil {
+			return nil, err
+		}
+		for kk, vv := range resources {
+			r[kk] = vv
+		}
 	}
-
 	return r, nil
 }
 
 // Show renders a Jsonnet dashboard as JSON, consuming a jsonnet filename
 func Show(config Config, jsonnetFile string, targets []string) error {
-	res, err := parse(jsonnetFile)
+	resources, err := parse(config, jsonnetFile)
 	if err != nil {
 		return err
 	}
 
-	if interactive && len(res) >= 2 {
-		var items []term.PageItem
-		for _, r := range res {
+	var items []term.PageItem
+	for _, resource := range resources {
+		handler := resource.Handler
+		resource = *handler.Unprepare(resource)
+
+		rep, err := resource.GetRepresentation()
+		if err != nil {
+			return err
+		}
+		if interactive {
 			items = append(items, term.PageItem{
-				Name:    fmt.Sprintf("%s/%s", r.Kind(), r.UID()),
-				Content: mustYAML(r),
+				Name:    fmt.Sprintf("%s/%s", resource.Kind(), resource.UID),
+				Content: rep,
 			})
+		} else {
+			fmt.Printf("%s/%s:\n", resource.Kind(), resource.UID)
+			fmt.Println(rep)
 		}
+	}
+	if interactive {
 		return term.Page(items)
 	}
-
-	fmt.Print(res.String())
 	return nil
 }
 
-func mustYAML(i interface{}) string {
-	data, err := yaml.Marshal(i)
-	if err != nil {
-		panic(err)
-	}
-	return string(data)
-}
-
-// Diff renders a Jsonnet dashboard and compares it with what is found in Grafana
+// Diff renders Jsonnet resources and compares them to those at the endpoints
 func Diff(config Config, jsonnetFile string, targets []string) error {
-	boards, err := renderDashboards(jsonnetFile, targets, 0)
+	resources, err := parse(config, jsonnetFile)
 	if err != nil {
 		return err
 	}
 
-	for _, board := range boards {
-		uid := board.UID()
-		existingBoard, err := getDashboard(config, board.UID())
+	for _, resource := range resources {
+		handler := resource.Handler
+		local, err := resource.GetRepresentation()
+		if err != nil {
+			return nil
+		}
+		resource = *handler.Unprepare(resource)
+		uid := resource.UID
+		remote, err := handler.GetRemote(resource.UID)
 		if err == ErrNotFound {
-			log.Println(uid, yellow("not present in Grafana"))
+			log.Printf("%s/%s %s\n", resource.Path, uid, Yellow("not present in "+resource.Kind()))
 			continue
 		}
 		if err != nil {
-			return fmt.Errorf("Error retrieving dashboard %s: %v", uid, err)
+			return fmt.Errorf("Error retrieving resource from %s %s: %v", resource.Kind(), uid, err)
+		}
+		remote = handler.Unprepare(*remote)
+		remoteRepresentation, err := (*remote).GetRepresentation()
```

Comment:
definitely! It would just wrap the existing functions though.

Reply:
This whole block would move into the Handler then, making the 'workflow' code much easier to understand. This is what I meant with moving the prepare/unprepare into the Handler, it might be wrapper functions, but at least the workflow is about the workflow, not about the logic. It prevents the need for public prepare/unprepare functions. It might result in a bit of code duplication, but it is specific to the handler whether a prep/unprep is needed, it is not a concern of the workflow.